lib/Conversion/SeqToSV/FirRegLowering.cpp

```diff
@@ -20,34 +20,64 @@ using llvm::MapVector;
-                      llvm::DenseSet<Operation *> &forwardSlice,
-                      llvm::function_ref<bool(Operation *)> filter = nullptr) {
-  SmallVector<Operation *> worklist({root});
-
-  while (!worklist.empty()) {
-    Operation *op = worklist.pop_back_val();
-
-    if (!op)
-      continue;
-
-    if (filter && !filter(op))
-      continue;
+bool ReachableMuxes::isMuxReachableFrom(seq::FirRegOp regOp,
+                                        comb::MuxOp muxOp) {
+  return llvm::any_of(regOp.getResult().getUsers(), [&](Operation *user) {
+    if (opBlocksReachability(user))
+      return false;
+    buildReachabilityFrom(user);
+    return reachableMuxes[user].contains(muxOp);
+  });
+}
 
-    for (Region &region : op->getRegions())
-      for (Block &block : region)
-        for (Operation &blockOp : block)
-          if (forwardSlice.insert(&blockOp).second)
-            worklist.push_back(&blockOp);
-    for (Value result : op->getResults())
-      for (Operation *userOp : result.getUsers())
-        if (forwardSlice.insert(userOp).second)
-          worklist.push_back(userOp);
-
-    forwardSlice.insert(op);
+void ReachableMuxes::buildReachabilityFrom(Operation *startNode) {
+  // This is a backward dataflow analysis.
+  // First build a graph rooted at the `startNode`. Every user of an operation
+  // that doesnot block the reachability is a child node. Then, the ops that
+  // are reachable from a node is computed as the union of the Reachability of
+  // all its child nodes.
+  // The dataflow can be expressed as, for all child in the Children(node)
+  // Reachability(node) = node + Union{Reachability(child)}
+  if (visited.find(startNode) != visited.end())
+    return;
+  // The stack to record enough information for an iterative post-order
+  // traversal.
+
+  llvm::SmallVector<OpUserInfo> stk;
+
+  stk.emplace_back(startNode);
+
+  while (!stk.empty()) {
+    auto &info = stk.back();
+    Operation *currentNode = info.op;
+
+    // Node is being visited for the first time.
+    if (info.userIter == info.userRange.begin())
+      visited.insert(currentNode);
+    if (info.getNextValid(info.userIter)) {
+      Operation *child = *info.userIter;
+      ++info.userIter;
+      if (visited.find(child) == visited.end())
+        stk.emplace_back(child);
+
+    } else { // All children of the node have been visited
+      // Any op is reachable from itself.
+      reachableMuxes[currentNode].insert(currentNode);
+      auto userIterator = info.userRange.begin();
+      while (info.getNextValid(userIterator)) {
+        Operation *childOp = *userIterator;
+        reachableMuxes[currentNode].insert(childOp);
+        // Propagate the reachability backwards from m to currentNode.
+        auto iter = reachableMuxes.find(childOp);
+
+        if (iter != reachableMuxes.end())
+          reachableMuxes[currentNode].insert(iter->getSecond().begin(),
+                                             iter->getSecond().end());
```

Comment:
The propagation of reachability sets -- does this work in the presence of cycles?

(where the reachable set for each node in the cycle should be the same, with all nodes in the cycle plus any children not part of the cycle)

I think maybe the idea is that cycles are not possible here, is that right?

(that's why can do the dataflow union thing in a single "pass" (on the return of DFS'ing through it)?)

Reply:
You are right, this implementation assumes absence of comb cycles. This PR actually implements the check: "Is a mux reachable from a register, given that the register is reachable from the register"